File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 154540. 무인도 여행/kotlin/Solution.kt
import java.util.*

class Solution {
    private val dr = intArrayOf(0, 1, 0, -1)
    private val dc = intArrayOf(1, 0, -1, 0)

    fun solution(maps: Array<String>): IntArray {
        val width = maps[0].length
        val height = maps.size
        val board = Array(height) { r -> IntArray(width) { c ->
            if (maps[r][c] == 'X') 0 else maps[r][c].code - 48 }
        }

        var sum = 0
        fun dfs(r: Int, c: Int) {
            var nr: Int
            var nc: Int
            for (i in 0 until 4) {
                nr = r + dr[i]
                nc = c + dc[i]
                if ((nr in 0 until height) && (nc in 0 until width) && board[nr][nc] > 0) {
                    sum += board[nr][nc]
                    board[nr][nc] = 0
                    dfs(nr, nc)
                }
            }
        }

        val answer = ArrayList<Int>()

        for (r in 0 until height) {
            for (c in 0 until width) {
                if (board[r][c] > 0) {
                    sum = board[r][c]
                    board[r][c] = 0
                    dfs(r, c)
                    answer += sum
                }
            }
        }
        if (answer.isEmpty()) {
            return intArrayOf(-1)
        }
        else {
            answer.sort()
            return answer.toIntArray()
        }
    }
}

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 154540. 무인도 여행/python/solution.py
directions = [(1, 0), (0, 1), (-1, 0), (0, -1)]


def solution(maps):
    maps = [list(map) for map in maps]
    result = []

    def dfs(r, c, maps):
        stack = [(r, c)]
        day = int(maps[r][c])
        maps[r][c] = 'X'

        while stack:
            cr, cc = stack.pop()
            for dr, dc in directions:
                nr, nc = cr + dr, cc + dc

                if 0 <= nr < len(maps) and 0 <= nc < len(maps[0]) and maps[nr][nc] != 'X':
                    day += int(maps[nr][nc])
                    maps[nr][nc] = 'X'
                    stack.append((nr, nc))
        return day

    for r in range(len(maps)):
        for c in range(len(maps[0])):
            if maps[r][c] != 'X':
                day = dfs(r, c, maps)
                result.append(day)
    return [-1] if not result else sorted(result)
